### PR TITLE
Use actualy dict for dict_squash and dict_copy from segment

### PR DESCRIPTION
**Pull request by ClementWalter**

### CODE REVIEW

File: cairo/src/state.cairo [hunk 1]
```diff
@@ -417,6 +414,30 @@ namespace Internals {
...
+        accounts_start: DictAccess*, accounts_end: DictAccess*
+    ) {
+        if (accounts_start == accounts_end) {
+            return ();
+        }
+
+        if (accounts_start.new_value == 0) {
+            // If we do a dict_read on an unexisting account, `prev_value` and `new_value` are set to 0.
+            // However we expected pointers to model.Account, and casting 0 to model.Account* will
+            // cause a "Memory address must be relocatable" error.
```
Comment (obatirou) [inline cairo/src/state.cairo:430]: Comment needs fixing

File: cairo/src/utils/dict.cairo [hunk 2]
```diff
@@ -1,134 +1,42 @@
...
-    static_assert len == [ap - 2];
-    static_assert dict == [ap - 1];
-
-    jmp loop if len != 0;
-
-    return (values_len, values);
+    let (local new_start: DictAccess*) = alloc();
+    let new_end = new_start + (dict_end - dict_start);
+    memcpy(new_start, dict_start, dict_end - dict_start);
+    %{ dict_copy %}
```
Comment (obatirou) [inline cairo/src/utils/dict.cairo:16]: Why this hint ?

File: cairo/src/utils/dict.cairo [hunk 3]
```diff
@@ -1,134 +1,42 @@
...
-
-    static_assert squashed_start == [ap - 3];
-    static_assert dict_len == [ap - 2];
-    static_assert new_ptr == [ap - 1];
-
-    jmp loop if dict_len != 0;
 
-    return (new_start, new_ptr);
+    %{ dict_squash %}
+    ap += 1;
```
Comment (obatirou) [inline cairo/src/utils/dict.cairo:27]: From my understanding this si due to the hint doing  `memory[ap] = base`.
What happen if the wrong value is filled ?
There is no contraints on this memory cell unless I miss something

File: cairo/src/state.cairo (hunk 1 above)
Comment (ClementWalter) [reply cairo/src/state.cairo:430]: ? what is unclear (same comment as in kakarot btw)

File: cairo/src/utils/dict.cairo (hunk 2 above)
Comment (ClementWalter) [reply cairo/src/utils/dict.cairo:16]: the hint is for registering the dict into the underlying `DictManager`.

File: cairo/src/state.cairo (hunk 1 above)
Comment (obatirou) [reply cairo/src/state.cairo:430]: mmh it seems I hallucinated an other comment. Sorry for that. It's clear

File: cairo/src/utils/dict.cairo (hunk 3 above)
Comment (obatirou) [reply cairo/src/utils/dict.cairo:28]: I did miss something 
Thanks for the clarification

**Review by obatirou — APPROVED** [review]

**Pull request merged**